packages/ampd-sdk/src/grpc/connection_manager.rs

```diff
@@ -0,0 +1,257 @@
+use ampd::url::Url;
+use error_stack::{Result, ResultExt as _};
+use report::ResultExt;
+use tokio::sync::{mpsc, watch};
+use tokio::time::Duration;
+use tonic::transport;
+use tracing::{error, info, warn};
+
+use crate::future::{with_retry, RetryPolicy};
+use crate::grpc::error::{AppError, Error};
+use crate::grpc::utils::{ClientMessage, ConnectionHandle, ConnectionState};
+
+// TODO: make these configurable
+const DEFAULT_RETRY_POLICY: RetryPolicy = RetryPolicy::RepeatConstant {
+    sleep: Duration::from_millis(100),
+    max_attempts: 3,
+};
+const DEFAULT_INITIAL_TIMEOUT: Duration = Duration::from_secs(3);
+const DEFAULT_RPC_TIMEOUT: Duration = Duration::from_secs(5);
+const KEEPALIVE_TIME: Duration = Duration::from_secs(30);
+const KEEPALIVE_TIMEOUT: Duration = Duration::from_secs(3);
+const KEEPALIVE_WHILE_IDLE: bool = true;
+
+/// Connection manager that handles connection state and retries.
+///
+/// This is a multi-producer single consumer pattern:
+/// - The manager maintains the connection state via a watch channel
+/// - Multiple clients can subscribe to connection state changes
+/// - Clients communicate back to the manager via mpsc channel
+/// - The manager handles connection retries and failure recovery
+#[derive(Debug)]
+pub struct ConnectionManager {
```

Comment:
event though it's currently a pool of one, I'd prefer the name `ConnectionPool`. We already have too many handlers, processors, and managers. Let's be more specific when we can